src/Http/WebUtilities/src/FileBufferingPipeWriter.cs

```diff
@@ -0,0 +1,164 @@
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    // TODO: Implement spooling to disk
+    // For perf it would be nice to write straight to the pipe until a threshold, to optimize performance for small payloads.
+    // Later switch to a file stream approach, probably keeping a buffer to write in large chunks to disk.
+    // This approach would be faster for small write but larger memory usage for big ones, so next will require to find new thresholds.
+    // However it would probbaly not meet the buffering requirement.
+    /// <summary>
+    /// A <see cref="Stream"/> that buffers content to be written to disk.
+    /// </summary>
+    public sealed class FileBufferingPipeWriter : PipeWriter, IAsyncDisposable
+    {
+        private const int DefaultMemoryThreshold = 32 * 1024; // 32k
+        private readonly PipeWriter _pipeWriter;
+        private readonly int _memoryThreshold;
+        private readonly long? _bufferLimit;
+        private readonly Func<string> _tempFileDirectoryAccessor;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="FileBufferingWriteStream"/>.
+        /// </summary>
+        /// <param name="pipeWriter"></param>
+        /// <param name="memoryThreshold">
+        /// The maximum amount of memory in bytes to allocate before switching to a file on disk.
+        /// Defaults to 32kb.
+        /// </param>
+        /// <param name="bufferLimit">
+        /// The maximum amount of bytes that the <see cref="FileBufferingWriteStream"/> is allowed to buffer.
+        /// </param>
+        /// <param name="tempFileDirectoryAccessor">Provides the location of the directory to write buffered contents to.
+        /// When unspecified, uses the value specified by the environment variable <c>ASPNETCORE_TEMP</c> if available, otherwise
+        /// uses the value returned by <see cref="Path.GetTempPath"/>.
+        /// </param>
+        public FileBufferingPipeWriter(
+            PipeWriter pipeWriter,
+            int memoryThreshold = DefaultMemoryThreshold,
+            long? bufferLimit = null,
+            Func<string>? tempFileDirectoryAccessor = null)
+        {
+            if (memoryThreshold < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(memoryThreshold));
+            }
+
+            if (bufferLimit != null && bufferLimit < memoryThreshold)
+            {
+                // We would expect a limit at least as much as memoryThreshold
+                throw new ArgumentOutOfRangeException(nameof(bufferLimit), $"{nameof(bufferLimit)} must be larger than {nameof(memoryThreshold)}.");
+            }
+
+            _pipeWriter = pipeWriter;
+            _memoryThreshold = memoryThreshold;
+            _bufferLimit = bufferLimit;
+            _tempFileDirectoryAccessor = tempFileDirectoryAccessor ?? AspNetCoreTempDirectory.TempDirectoryFactory;
+            PagedByteBuffer = new PagedByteBuffer(ArrayPool<byte>.Shared);
```

Comment:
I'm not sure we should be using this but now I'm convinced we should expose something in Pipelines for this.

Reply:
Well this is not actually used here, I just brought it from the other implementation as I thought I would have needed it later to avoid chunking. Based on the outcome of the that conversation, this can hopefully be avoided.

Reply:
Hey @davidfowl , any progress on exposing something in Pipelines to support this?